src/main/java/oliver/Ui.java

```diff
@@ -0,0 +1,165 @@
+            builder.deleteCharAt(builder.length() - 1);
+            return "Here are the tasks in your list:\n" + builder;
+        }
+    }
+
+    /**
+     * Shows the confirmation message when user marks a task as done.
+     *
+     * @param list the list of tasks
+     * @param index the index of the task user marks as done
+     * @return string representation of the confirmation message
+     */
+    public String showMarked(TaskList list, int index) {
+        list.get(index).markAsDone();
+        return "Nice! I've marked this task as done:\n" + list.get(index);
+    }
+
+    /**
+     * Shows the confirmation message when user marks a task as undone.
+     *
+     * @param list the list of tasks
+     * @param index the index of the task user marks as undone
+     * @return string representation of the confirmation message
+     */
+    public String showUnmarked(TaskList list, int index) {
+        list.get(index).markAsUndone();
+        return "Ok, I've marked this task as not done yet:\n" + list.get(index);
+    }
+
+    /**
+     * Shows the matching results of user's search.
+     *
+     * @param list the list of tasks
+     * @param keyword the word to be searched for in the list of tasks
+     * @return string representation of the results
+     */
+    public String showSearch(TaskList list, String keyword) {
+        if (list.isEmpty()) {
+            return "There are no tasks in your list.";
+        } else {
+            TaskList filteredList = list.filter(keyword);
+            if (filteredList.isEmpty()) {
+                return "There are no matching tasks in your list.";
+            } else {
+                StringBuilder builder = new StringBuilder();
+                for (int i = 1; i <= filteredList.getSize(); i++) {
+                    builder.append(String.format("%d. %s\n", i, filteredList.get(i - 1)));
+                }
+                // Remove the last newline char
+                builder.deleteCharAt(builder.length() - 1);
+                return "Here are the matching tasks in your list:\n" + builder;
+            }
+        }
+    }
+
+    /**
+     * Shows the error message when arguments are missing for a command.
+     *
+     * @return string representation of the error message
+     */
+    public String showMissingArgsError() {
+        return "Missing arguments for this command.";
+    }
+
+    /**
+     * Shows the error message when invalid arguments are provided for a command.
+     *
+     * @return string representation of the error message
+     */
+    public String showInvalidArgsError() {
+        return "Invalid arguments provided for this command.";
+    }
+
+    /**
+     * Shows the error message when arguments provided for a command are out of range.
+     *
+     * @return string representation of the error message
+     */
+    public String showOutOfRangeError() {
+        return "No such task exists. Task number out of range.";
+    }
+}
```

Comment:
Perhaps we could use a custom exception class to handle all of these, which may help simplify the logic of your Ui class.